// firebase-admin/lib/utils/error.js
/*! firebase-admin v9.12.0 */
"use strict";
/*!
 * @license
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceIdClientErrorCode = exports.InstallationsClientErrorCode = exports.MessagingClientErrorCode = exports.AuthClientErrorCode = exports.AppErrorCodes = exports.FirebaseProjectManagementError = exports.FirebaseMessagingError = exports.FirebaseInstallationsError = exports.FirebaseInstanceIdError = exports.FirebaseFirestoreError = exports.FirebaseDatabaseError = exports.FirebaseAuthError = exports.FirebaseAppError = exports.PrefixedFirebaseError = exports.FirebaseError = void 0;
var deep_copy_1 = require("../utils/deep-copy");
/**
 * Firebase error code structure. This extends Error.
 *
 * @param {ErrorInfo} errorInfo The error information (code and message).
 * @constructor
 */
var FirebaseError = /** @class */ (function (_super) {
    __extends(FirebaseError, _super);
    function FirebaseError(errorInfo) {
        var _this = _super.call(this, errorInfo.message) || this;
        _this.errorInfo = errorInfo;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = FirebaseError.prototype;
        return _this;
    }
    Object.defineProperty(FirebaseError.prototype, "code", {
        /** @return {string} The error code. */
        get: function () {
            return this.errorInfo.code;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(FirebaseError.prototype, "message", {
        /** @return {string} The error message. */
        get: function () {
            return this.errorInfo.message;
        },
        enumerable: false,
        configurable: true
    });
    /** @return {object} The object representation of the error. */
    FirebaseError.prototype.toJSON = function () {
        return {
            code: this.code,
            message: this.message,
        };
    };
    return FirebaseError;
}(Error));
exports.FirebaseError = FirebaseError;
/**
 * A FirebaseError with a prefix in front of the error code.
 *
 * @param {string} codePrefix The prefix to apply to the error code.
 * @param {string} code The error code.
 * @param {string} message The error message.
 * @constructor
 */
var PrefixedFirebaseError = /** @class */ (function (_super) {
    __extends(PrefixedFirebaseError, _super);
    function PrefixedFirebaseError(codePrefix, code, message) {
        var _this = _super.call(this, {
            code: codePrefix + "/" + code,
            message: message,
        }) || this;
        _this.codePrefix = codePrefix;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = PrefixedFirebaseError.prototype;
        return _this;
    }
    /**
     * Allows the error type to be checked without needing to know implementation details
     * of the code prefixing.
     *
     * @param {string} code The non-prefixed error code to test against.
     * @return {boolean} True if the code matches, false otherwise.
     */
    PrefixedFirebaseError.prototype.hasCode = function (code) {
        return this.codePrefix + "/" + code === this.code;
    };
    return PrefixedFirebaseError;
}(FirebaseError));
exports.PrefixedFirebaseError = PrefixedFirebaseError;
/**
 * Firebase App error code structure. This extends PrefixedFirebaseError.
 *
 * @param {string} code The error code.
 * @param {string} message The error message.
 * @constructor
 */
var FirebaseAppError = /** @class */ (function (_super) {
    __extends(FirebaseAppError, _super);
    function FirebaseAppError(code, message) {
        var _this = _super.call(this, 'app', code, message) || this;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = FirebaseAppError.prototype;
        return _this;
    }
    return FirebaseAppError;
}(PrefixedFirebaseError));
exports.FirebaseAppError = FirebaseAppError;
/**
 * Firebase Auth error code structure. This extends PrefixedFirebaseError.
 *
 * @param {ErrorInfo} info The error code info.
 * @param {string} [message] The error message. This will override the default
 *     message if provided.
 * @constructor
 */
var FirebaseAuthError = /** @class */ (function (_super) {
    __extends(FirebaseAuthError, _super);
    function FirebaseAuthError(info, message) {
        var _this = 
        // Override default message if custom message provided.
        _super.call(this, 'auth', info.code, message || info.message) || this;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = FirebaseAuthError.prototype;
        return _this;
    }
    /**
     * Creates the developer-facing error corresponding to the backend error code.
     *
     * @param {string} serverErrorCode The server error code.
     * @param {string} [message] The error message. The default message is used
     *     if not provided.
     * @param {object} [rawServerResponse] The error's raw server response.
     * @return {FirebaseAuthError} The corresponding developer-facing error.
     */
    FirebaseAuthError.fromServerError = function (serverErrorCode, message, rawServerResponse) {
        // serverErrorCode could contain additional details:
        // ERROR_CODE : Detailed message which can also contain colons
        var colonSeparator = (serverErrorCode || '').indexOf(':');
        var customMessage = null;
        if (colonSeparator !== -1) {
            customMessage = serverErrorCode.substring(colonSeparator + 1).trim();
            serverErrorCode = serverErrorCode.substring(0, colonSeparator).trim();
        }
        // If not found, default to internal error.
        var clientCodeKey = AUTH_SERVER_TO_CLIENT_CODE[serverErrorCode] || 'INTERNAL_ERROR';
        var error = deep_copy_1.deepCopy(AuthClientErrorCode[clientCodeKey]);
        // Server detailed message should have highest priority.
        error.message = customMessage || message || error.message;
        if (clientCodeKey === 'INTERNAL_ERROR' && typeof rawServerResponse !== 'undefined') {
            try {
                error.message += " Raw server response: \"" + JSON.stringify(rawServerResponse) + "\"";
            }
            catch (e) {
                // Ignore JSON parsing error.
            }
        }
        return new FirebaseAuthError(error);
    };
    return FirebaseAuthError;
}(PrefixedFirebaseError));
exports.FirebaseAuthError = FirebaseAuthError;
/**
 * Firebase Database error code structure. This extends FirebaseError.
 *
 * @param {ErrorInfo} info The error code info.
 * @param {string} [message] The error message. This will override the default
 *     message if provided.
 * @constructor
 */
var FirebaseDatabaseError = /** @class */ (function (_super) {
    __extends(FirebaseDatabaseError, _super);
    function FirebaseDatabaseError(info, message) {
        // Override default message if custom message provided.
        return _super.call(this, { code: 'database/' + info.code, message: message || info.message }) || this;
    }
    return FirebaseDatabaseError;
}(FirebaseError));
exports.FirebaseDatabaseError = FirebaseDatabaseError;
/**
 * Firebase Firestore error code structure. This extends FirebaseError.
 *
 * @param {ErrorInfo} info The error code info.
 * @param {string} [message] The error message. This will override the default
 *     message if provided.
 * @constructor
 */
var FirebaseFirestoreError = /** @class */ (function (_super) {
    __extends(FirebaseFirestoreError, _super);
    function FirebaseFirestoreError(info, message) {
        // Override default message if custom message provided.
        return _super.call(this, { code: 'firestore/' + info.code, message: message || info.message }) || this;
    }
    return FirebaseFirestoreError;
}(FirebaseError));
exports.FirebaseFirestoreError = FirebaseFirestoreError;
/**
 * Firebase instance ID error code structure. This extends FirebaseError.
 *
 * @param {ErrorInfo} info The error code info.
 * @param {string} [message] The error message. This will override the default
 *     message if provided.
 * @constructor
 */
var FirebaseInstanceIdError = /** @class */ (function (_super) {
    __extends(FirebaseInstanceIdError, _super);
    function FirebaseInstanceIdError(info, message) {
        var _this = 
        // Override default message if custom message provided.
        _super.call(this, { code: 'instance-id/' + info.code, message: message || info.message }) || this;
        _this.__proto__ = FirebaseInstanceIdError.prototype;
        return _this;
    }
    return FirebaseInstanceIdError;
}(FirebaseError));
exports.FirebaseInstanceIdError = FirebaseInstanceIdError;
/**
 * Firebase Installations service error code structure. This extends `FirebaseError`.
 *
 * @param info The error code info.
 * @param message The error message. This will override the default
 *     message if provided.
 * @constructor
 */
var FirebaseInstallationsError = /** @class */ (function (_super) {
    __extends(FirebaseInstallationsError, _super);
    function FirebaseInstallationsError(info, message) {
        var _this = 
        // Override default message if custom message provided.
        _super.call(this, { code: 'installations/' + info.code, message: message || info.message }) || this;
        _this.__proto__ = FirebaseInstallationsError.prototype;
        return _this;
    }
    return FirebaseInstallationsError;
}(FirebaseError));
exports.FirebaseInstallationsError = FirebaseInstallationsError;
/**
 * Firebase Messaging error code structure. This extends PrefixedFirebaseError.
 *
 * @param {ErrorInfo} info The error code info.
 * @param {string} [message] The error message. This will override the default message if provided.
 * @constructor
 */
var FirebaseMessagingError = /** @class */ (function (_super) {
    __extends(FirebaseMessagingError, _super);
    function FirebaseMessagingError(info, message) {
        var _this = 
        // Override default message if custom message provided.
        _super.call(this, 'messaging', info.code, message || info.message) || this;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = FirebaseMessagingError.prototype;
        return _this;
    }
    /**
     * Creates the developer-facing error corresponding to the backend error code.
     *
     * @param {string} serverErrorCode The server error code.
     * @param {string} [message] The error message. The default message is used
     *     if not provided.
     * @param {object} [rawServerResponse] The error's raw server response.
     * @return {FirebaseMessagingError} The corresponding developer-facing error.
     */
    FirebaseMessagingError.fromServerError = function (serverErrorCode, message, rawServerResponse) {
        // If not found, default to unknown error.
        var clientCodeKey = 'UNKNOWN_ERROR';
        if (serverErrorCode && serverErrorCode in MESSAGING_SERVER_TO_CLIENT_CODE) {
            clientCodeKey = MESSAGING_SERVER_TO_CLIENT_CODE[serverErrorCode];
        }
        var error = deep_copy_1.deepCopy(MessagingClientErrorCode[clientCodeKey]);
        error.message = message || error.message;
        if (clientCodeKey === 'UNKNOWN_ERROR' && typeof rawServerResponse !== 'undefined') {
            try {
                error.message += " Raw server response: \"" + JSON.stringify(rawServerResponse) + "\"";
            }
            catch (e) {
                // Ignore JSON parsing error.
            }
        }
        return new FirebaseMessagingError(error);
    };
    FirebaseMessagingError.fromTopicManagementServerError = function (serverErrorCode, message, rawServerResponse) {
        // If not found, default to unknown error.
        var clientCodeKey = TOPIC_MGT_SERVER_TO_CLIENT_CODE[serverErrorCode] || 'UNKNOWN_ERROR';
        var error = deep_copy_1.deepCopy(MessagingClientErrorCode[clientCodeKey]);
        error.message = message || error.message;
        if (clientCodeKey === 'UNKNOWN_ERROR' && typeof rawServerResponse !== 'undefined') {
            try {
                error.message += " Raw server response: \"" + JSON.stringify(rawServerResponse) + "\"";
            }
            catch (e) {
                // Ignore JSON parsing error.
            }
        }
        return new FirebaseMessagingError(error);
    };
    return FirebaseMessagingError;
}(PrefixedFirebaseError));
exports.FirebaseMessagingError = FirebaseMessagingError;
/**
 * Firebase project management error code structure. This extends PrefixedFirebaseError.
 *
 * @param {ProjectManagementErrorCode} code The error code.
 * @param {string} message The error message.
 * @constructor
 */
var FirebaseProjectManagementError = /** @class */ (function (_super) {
    __extends(FirebaseProjectManagementError, _super);
    function FirebaseProjectManagementError(code, message) {
        var _this = _super.call(this, 'project-management', code, message) || this;
        /* tslint:disable:max-line-length */
        // Set the prototype explicitly. See the following link for more details:
        // https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work
        /* tslint:enable:max-line-length */
        _this.__proto__ = FirebaseProjectManagementError.prototype;
        return _this;
    }
    return FirebaseProjectManagementError;
}(PrefixedFirebaseError));
exports.FirebaseProjectManagementError = FirebaseProjectManagementError;
/**
 * App client error codes and their default messages.
 */
var AppErrorCodes = /** @class */ (function () {
    function AppErrorCodes() {
    }
    AppErrorCodes.APP_DELETED = 'app-deleted';
    AppErrorCodes.DUPLICATE_APP = 'duplicate-app';
    AppErrorCodes.INVALID_ARGUMENT = 'invalid-argument';
    AppErrorCodes.INTERNAL_ERROR = 'internal-error';
    AppErrorCodes.INVALID_APP_NAME = 'invalid-app-name';
    AppErrorCodes.INVALID_APP_OPTIONS = 'invalid-app-options';
    AppErrorCodes.INVALID_CREDENTIAL = 'invalid-credential';
    AppErrorCodes.NETWORK_ERROR = 'network-error';
    AppErrorCodes.NETWORK_TIMEOUT = 'network-timeout';
    AppErrorCodes.NO_APP = 'no-app';
    AppErrorCodes.UNABLE_TO_PARSE_RESPONSE = 'unable-to-parse-response';
    return AppErrorCodes;
}());
exports.AppErrorCodes = AppErrorCodes;
/**
 * Auth client error codes and their default messages.
 */
var AuthClientErrorCode = /** @class */ (function () {
    function AuthClientErrorCode() {
    }
    AuthClientErrorCode.BILLING_NOT_ENABLED = {
        code: 'billing-not-enabled',
        message: 'Feature requires billing to be enabled.',
    };
    AuthClientErrorCode.CLAIMS_TOO_LARGE = {
        code: 'claims-too-large',
        message: 'Developer claims maximum payload size exceeded.',
    };
    AuthClientErrorCode.CONFIGURATION_EXISTS = {
        code: 'configuration-exists',
        message: 'A configuration already exists with the provided identifier.',
    };
    AuthClientErrorCode.CONFIGURATION_NOT_FOUND = {
        code: 'configuration-not-found',
        message: 'There is no configuration corresponding to the provided identifier.',
    };
    AuthClientErrorCode.ID_TOKEN_EXPIRED = {
        code: 'id-token-expired',
        message: 'The provided Firebase ID token is expired.',
    };
    AuthClientErrorCode.INVALID_ARGUMENT = {
        code: 'argument-error',
        message: 'Invalid argument provided.',
    };
    AuthClientErrorCode.INVALID_CONFIG = {
        code: 'invalid-config',
        message: 'The provided configuration is invalid.',
    };
    AuthClientErrorCode.EMAIL_ALREADY_EXISTS = {
        code: 'email-already-exists',
        message: 'The email address is already in use by another account.',
    };
    AuthClientErrorCode.EMAIL_NOT_FOUND = {
        code: 'email-not-found',
        message: 'There is no user record corresponding to the provided email.',
    };
    AuthClientErrorCode.FORBIDDEN_CLAIM = {
        code: 'reserved-claim',
        message: 'The specified developer claim is reserved and cannot be specified.',
    };
    AuthClientErrorCode.INVALID_ID_TOKEN = {
        code: 'invalid-id-token',
        message: 'The provided ID token is not a valid Firebase ID token.',
    };
    AuthClientErrorCode.ID_TOKEN_REVOKED = {
        code: 'id-token-revoked',
        message: 'The Firebase ID token has been revoked.',
    };
    AuthClientErrorCode.INTERNAL_ERROR = {
        code: 'internal-error',
        message: 'An internal error has occurred.',
    };
    AuthClientErrorCode.INVALID_CLAIMS = {
        code: 'invalid-claims',
        message: 'The provided custom claim attributes are invalid.',
    };
    AuthClientErrorCode.INVALID_CONTINUE_URI = {
        code: 'invalid-continue-uri',
        message: 'The continue URL must be a valid URL string.',
    };
    AuthClientErrorCode.INVALID_CREATION_TIME = {
        code: 'invalid-creation-time',
        message: 'The creation time must be a valid UTC date string.',
    };
    AuthClientErrorCode.INVALID_CREDENTIAL = {
        code: 'invalid-credential',
        message: 'Invalid credential object provided.',
    };
    AuthClientErrorCode.INVALID_DISABLED_FIELD = {
        code: 'invalid-disabled-field',
        message: 'The disabled field must be a boolean.',
    };
    AuthClientErrorCode.INVALID_DISPLAY_NAME = {
        code: 'invalid-display-name',
        message: 'The displayName field must be a valid string.',
    };
    AuthClientErrorCode.INVALID_DYNAMIC_LINK_DOMAIN = {
        code: 'invalid-dynamic-link-domain',
        message: 'The provided dynamic link domain is not configured or authorized ' +
            'for the current project.',
    };
    AuthClientErrorCode.INVALID_EMAIL_VERIFIED = {
        code: 'invalid-email-verified',
        message: 'The emailVerified field must be a boolean.',
    };
    AuthClientErrorCode.INVALID_EMAIL = {
        code: 'invalid-email',
        message: 'The email address is improperly formatted.',
    };
    AuthClientErrorCode.INVALID_ENROLLED_FACTORS = {
        code: 'invalid-enrolled-factors',
        message: 'The enrolled factors must be a valid array of MultiFactorInfo objects.',
    };
    AuthClientErrorCode.INVALID_ENROLLMENT_TIME = {
        code: 'invalid-enrollment-time',
        message: 'The second factor enrollment time must be a valid UTC date string.',
    };
    AuthClientErrorCode.INVALID_HASH_ALGORITHM = {
        code: 'invalid-hash-algorithm',
        message: 'The hash algorithm must match one of the strings in the list of ' +
            'supported algorithms.',
    };
    AuthClientErrorCode.INVALID_HASH_BLOCK_SIZE = {
        code: 'invalid-hash-block-size',
        message: 'The hash block size must be a valid number.',
    };
    AuthClientErrorCode.INVALID_HASH_DERIVED_KEY_LENGTH = {
        code: 'invalid-hash-derived-key-length',
        message: 'The hash derived key length must be a valid number.',
    };
    AuthClientErrorCode.INVALID_HASH_KEY = {
        code: 'invalid-hash-key',
        message: 'The hash key must a valid byte buffer.',
    };
    AuthClientErrorCode.INVALID_HASH_MEMORY_COST = {
        code: 'invalid-hash-memory-cost',
        message: 'The hash memory cost must be a valid number.',
    };
    AuthClientErrorCode.INVALID_HASH_PARALLELIZATION = {
        code: 'invalid-hash-parallelization',
        message: 'The hash parallelization must be a valid number.',
    };
    AuthClientErrorCode.INVALID_HASH_ROUNDS = {
        code: 'invalid-hash-rounds',
        message: 'The hash rounds must be a valid number.',
    };
    AuthClientErrorCode.INVALID_HASH_SALT_SEPARATOR = {
        code: 'invalid-hash-salt-separator',
        message: 'The hashing algorithm salt separator field must be a valid byte buffer.',
    };
    AuthClientErrorCode.INVALID_LAST_SIGN_IN_TIME = {
        code: 'invalid-last-sign-in-time',
        message: 'The last sign-in time must be a valid UTC date string.',
    };
    AuthClientErrorCode.INVALID_NAME = {
        code: 'invalid-name',
        message: 'The resource name provided is invalid.',
    };
    AuthClientErrorCode.INVALID_OAUTH_CLIENT_ID = {
        code: 'invalid-oauth-client-id',
        message: 'The provided OAuth client ID is invalid.',
    };
    AuthClientErrorCode.INVALID_PAGE_TOKEN = {
        code: 'invalid-page-token',
        message: 'The page token must be a valid non-empty string.',
    };
    AuthClientErrorCode.INVALID_PASSWORD = {
        code: 'invalid-password',
        message: 'The password must be a string with at least 6 characters.',
    };
    AuthClientErrorCode.INVALID_PASSWORD_HASH = {
        code: 'invalid-password-hash',
        message: 'The password hash must be a valid byte buffer.',
    };
    AuthClientErrorCode.INVALID_PASSWORD_SALT = {
        code: 'invalid-password-salt',
        message: 'The password salt must be a valid byte buffer.',
    };
    AuthClientErrorCode.INVALID_PHONE_NUMBER = {
        code: 'invalid-phone-number',
        message: 'The phone number must be a non-empty E.164 standard compliant identifier ' +
            'string.',
    };
    AuthClientErrorCode.INVALID_PHOTO_URL = {
        code: 'invalid-photo-url',
        message: 'The photoURL field must be a valid URL.',
    };
    AuthClientErrorCode.INVALID_PROJECT_ID = {
        code: 'invalid-project-id',
        message: 'Invalid parent project. Either parent project doesn\'t exist or didn\'t enable multi-tenancy.',
    };
    AuthClientErrorCode.INVALID_PROVIDER_DATA = {
        code: 'invalid-provider-data',
        message: 'The providerData must be a valid array of UserInfo objects.',
    };
    AuthClientErrorCode.INVALID_PROVIDER_ID = {
        code: 'invalid-provider-id',
        message: 'The providerId must be a valid supported provider identifier string.',
    };
    AuthClientErrorCode.INVALID_PROVIDER_UID = {
        code: 'invalid-provider-uid',
        message: 'The providerUid must be a valid provider uid string.',
    };
    AuthClientErrorCode.INVALID_OAUTH_RESPONSETYPE = {
        code: 'invalid-oauth-responsetype',
        message: 'Only exactly one OAuth responseType should be set to true.',
    };
    AuthClientErrorCode.INVALID_SESSION_COOKIE_DURATION = {
        code: 'invalid-session-cookie-duration',
        message: 'The session cookie duration must be a valid number in milliseconds ' +
            'between 5 minutes and 2 weeks.',
    };
    AuthClientErrorCode.INVALID_TENANT_ID = {
        code: 'invalid-tenant-id',
        message: 'The tenant ID must be a valid non-empty string.',
    };
    AuthClientErrorCode.INVALID_TENANT_TYPE = {
        code: 'invalid-tenant-type',
        message: 'Tenant type must be either "full_service" or "lightweight".',
    };
    AuthClientErrorCode.INVALID_TESTING_PHONE_NUMBER = {
        code: 'invalid-testing-phone-number',
        message: 'Invalid testing phone number or invalid test code provided.',
    };
    AuthClientErrorCode.INVALID_UID = {
        code: 'invalid-uid',
        message: 'The uid must be a non-empty string with at most 128 characters.',
    };
    AuthClientErrorCode.INVALID_USER_IMPORT = {
        code: 'invalid-user-import',
        message: 'The user record to import is invalid.',
    };
    AuthClientErrorCode.INVALID_TOKENS_VALID_AFTER_TIME = {
        code: 'invalid-tokens-valid-after-time',
        message: 'The tokensValidAfterTime must be a valid UTC number in seconds.',
    };
    AuthClientErrorCode.MISMATCHING_TENANT_ID = {
        code: 'mismatching-tenant-id',
        message: 'User tenant ID does not match with the current TenantAwareAuth tenant ID.',
    };
    AuthClientErrorCode.MISSING_ANDROID_PACKAGE_NAME = {
        code: 'missing-android-pkg-name',
        message: 'An Android Package Name must be provided if the Android App is ' +
            'required to be installed.',
    };
    AuthClientErrorCode.MISSING_CONFIG = {
        code: 'missing-config',
        message: 'The provided configuration is missing required attributes.',
    };
    AuthClientErrorCode.MISSING_CONTINUE_URI = {
        code: 'missing-continue-uri',
        message: 'A valid continue URL must be provided in the request.',
    };
    AuthClientErrorCode.MISSING_DISPLAY_NAME = {
        code: 'missing-display-name',
        message: 'The resource being created or edited is missing a valid display name.',
    };
    AuthClientErrorCode.MISSING_EMAIL = {
        code: 'missing-email',
        message: 'The email is required for the specified action. For example, a multi-factor user ' +
            'requires a verified email.',
    };
    AuthClientErrorCode.MISSING_IOS_BUNDLE_ID = {
        code: 'missing-ios-bundle-id',
        message: 'The request is missing an iOS Bundle ID.',
    };
    AuthClientErrorCode.MISSING_ISSUER = {
        code: 'missing-issuer',
        message: 'The OAuth/OIDC configuration issuer must not be empty.',
    };
    AuthClientErrorCode.MISSING_HASH_ALGORITHM = {
        code: 'missing-hash-algorithm',
        message: 'Importing users with password hashes requires that the hashing ' +
            'algorithm and its parameters be provided.',
    };
    AuthClientErrorCode.MISSING_OAUTH_CLIENT_ID = {
        code: 'missing-oauth-client-id',
        message: 'The OAuth/OIDC configuration client ID must not be empty.',
    };
    AuthClientErrorCode.MISSING_OAUTH_CLIENT_SECRET = {
        code: 'missing-oauth-client-secret',
        message: 'The OAuth configuration client secret is required to enable OIDC code flow.',
    };
    AuthClientErrorCode.MISSING_PROVIDER_ID = {
        code: 'missing-provider-id',
        message: 'A valid provider ID must be provided in the request.',
    };
    AuthClientErrorCode.MISSING_SAML_RELYING_PARTY_CONFIG = {
        code: 'missing-saml-relying-party-config',
        message: 'The SAML configuration provided is missing a relying party configuration.',
    };
    AuthClientErrorCode.MAXIMUM_TEST_PHONE_NUMBER_EXCEEDED = {
        code: 'test-phone-number-limit-exceeded',
        message: 'The maximum allowed number of test phone number / code pairs has been exceeded.',
    };
    AuthClientErrorCode.MAXIMUM_USER_COUNT_EXCEEDED = {
        code: 'maximum-user-count-exceeded',
        message: 'The maximum allowed number of users to import has been exceeded.',
    };
    AuthClientErrorCode.MISSING_UID = {
        code: 'missing-uid',
        message: 'A uid identifier is required for the current operation.',
    };
    AuthClientErrorCode.OPERATION_NOT_ALLOWED = {
        code: 'operation-not-allowed',
        message: 'The given sign-in provider is disabled for this Firebase project. ' +
            'Enable it in the Firebase console, under the sign-in method tab of the ' +
            'Auth section.',
    };
    AuthClientErrorCode.PHONE_NUMBER_ALREADY_EXISTS = {
        code: 'phone-number-already-exists',
        message: 'The user with the provided phone number already exists.',
    };
    AuthClientErrorCode.PROJECT_NOT_FOUND = {
        code: 'project-not-found',
        message: 'No Firebase project was found for the provided credential.',
    };
    AuthClientErrorCode.INSUFFICIENT_PERMISSION = {
        code: 'insufficient-permission',
        message: 'Credential implementation provided to initializeApp() via the "credential" property ' +
            'has insufficient permission to access the requested resource. See ' +
            'https://firebase.google.com/docs/admin/setup for details on how to authenticate this SDK ' +
            'with appropriate permissions.',
    };
    AuthClientErrorCode.QUOTA_EXCEEDED = {
        code: 'quota-exceeded',
        message: 'The project quota for the specified operation has been exceeded.',
    };
    AuthClientErrorCode.SECOND_FACTOR_LIMIT_EXCEEDED = {
        code: 'second-factor-limit-exceeded',
        message: 'The maximum number of allowed second factors on a user has been exceeded.',
    };
    AuthClientErrorCode.SECOND_FACTOR_UID_ALREADY_EXISTS = {
        code: 'second-factor-uid-already-exists',
        message: 'The specified second factor "uid" already exists.',
    };
    AuthClientErrorCode.SESSION_COOKIE_EXPIRED = {
        code: 'session-cookie-expired',
        message: 'The Firebase session cookie is expired.',
    };
    AuthClientErrorCode.SESSION_COOKIE_REVOKED = {
        code: 'session-cookie-revoked',
        message: 'The Firebase session cookie has been revoked.',
    };
    AuthClientErrorCode.TENANT_NOT_FOUND = {
        code: 'tenant-not-found',
        message: 'There is no tenant corresponding to the provided identifier.',
    };
    AuthClientErrorCode.UID_ALREADY_EXISTS = {
        code: 'uid-already-exists',
        message: 'The user with the provided uid already exists.',
    };
    AuthClientErrorCode.UNAUTHORIZED_DOMAIN = {
        code: 'unauthorized-continue-uri',
        message: 'The domain of the continue URL is not whitelisted. Whitelist the domain in the ' +
            'Firebase console.',
    };
    AuthClientErrorCode.UNSUPPORTED_FIRST_FACTOR = {
        code: 'unsupported-first-factor',
        message: 'A multi-factor user requires a supported first factor.',
    };
    AuthClientErrorCode.UNSUPPORTED_SECOND_FACTOR = {
        code: 'unsupported-second-factor',
        message: 'The request specified an unsupported type of second factor.',
    };
    AuthClientErrorCode.UNSUPPORTED_TENANT_OPERATION = {
        code: 'unsupported-tenant-operation',
        message: 'This operation is not supported in a multi-tenant context.',
    };
    AuthClientErrorCode.UNVERIFIED_EMAIL = {
        code: 'unverified-email',
        message: 'A verified email is required for the specified action. For example, a multi-factor user ' +
            'requires a verified email.',
    };
    AuthClientErrorCode.USER_NOT_FOUND = {
        code: 'user-not-found',
        message: 'There is no user record corresponding to the provided identifier.',
    };
    AuthClientErrorCode.NOT_FOUND = {
        code: 'not-found',
        message: 'The requested resource was
